spr/src/commit_message.rs

```diff
@@ -0,0 +1,426 @@
+// Write is needed by Command::Stdin.write_all()
+use std::io::Write;
+use std::process::{Command, Stdio};
+
+use crate::{
+    error::{Error, Result},
+};
+
+// Notice: use BTreeMap to make it easier to iterate trailer keys in order.
+pub type TrailerMap = std::collections::BTreeMap<String, Vec<String>>;
+
+#[derive(Debug, PartialEq)]
+pub struct CommitMessage {
+
+    /// Subject of the message (i.e. very first line)
+    pub subject:  String,
+
+    /// Body of the message, *EXCLUDING* the contents of the trailers
+    /// section. Empty string if no body.
+    pub body:     String,
+
+    /// Map of trailer keys to trailer values (e.g, "key: value...").
+    pub trailers: TrailerMap,
+}
+
+impl CommitMessage {
+
+    pub fn render(&self) -> String {
+        let mut ret: String = "".to_string();
```

Comment:
(pls feel free to ignore!)

possibly avoid the `mut` approach with something more functional?

```
   let subject = if ... { ... } else { "".to_string() }
   let body = if ...
   let trailers = if ...

   subject + body + trailers
```

Reply:
will keep in mind for next time, but will pass on this one (very low on b/w)